src/data/pnat/find.lean

```diff
@@ -0,0 +1,147 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky, Floris van Doorn
+-/
+import data.pnat.basic
+
+/-!
+# Explicit least witnesses to existentials on positive natural numbers
+
+Implemented via calling out to `nat.find`.
+
+-/
+
+namespace pnat
+
+variables {p q : ℕ+ → Prop} [decidable_pred p] [decidable_pred q] (h : ∃ n, p n)
+
+instance decidable_pred_exists_nat :
+  decidable_pred (λ n' : ℕ, ∃ (n : ℕ+) (hn : n' = n), p n)
+| 0       := is_false begin
+    push_neg,
+    rintro ⟨x, hx⟩,
+    simp [hx.ne]
+  end
+| (n' + 1) := if hp : p (n' + 1).to_pnat' then is_true ⟨(n' + 1).to_pnat', rfl, hp⟩
+  else is_false $ begin
+    contrapose! hp,
+    obtain ⟨n, hn, hp⟩ := hp,
+    simp [hn, hp]
+  end
+
+include h
+
+/--
+If `p` is a (decidable) predicate on `ℕ+` and `hp : ∃ (n : ℕ+), p n` is a proof that
+there exists some positive natural number satisfying `p`, then `pnat.find hp` is the
+smallest positive natural number satisfying `p`. Note that `pnat.find` is protected,
+meaning that you can't just write `find`, even if the `pnat` namespace is open.
+
+The API for `pnat.find` is:
+
+* `pnat.find_spec` is the proof that `pnat.find hp` satisfies `p`.
+* `pnat.find_min` is the proof that if `m < pnat.find hp` then `m` does not satisfy `p`.
+* `pnat.find_min'` is the proof that if `m` does satisfy `p` then `pnat.find hp ≤ m`.
+-/
+protected def find_aux
+  (H' : ∃ (n' : ℕ) (n : ℕ+) (hn' : n' = n), p n := exists.elim h (λ n hn, ⟨n, n, rfl, hn⟩)) : ℕ :=
+nat.find H'
+
+protected theorem find_aux_spec
+  (H' : ∃ (n' : ℕ) (n : ℕ+) (hn' : n' = n), p n := exists.elim h (λ n hn, ⟨n, n, rfl, hn⟩)) :
+  ∃ (n : ℕ+) (hn' : pnat.find_aux h = n), p n :=
+nat.find_spec H'
+
+/--
+If `p` is a (decidable) predicate on `ℕ+` and `hp : ∃ (n : ℕ+), p n` is a proof that
+there exists some positive natural number satisfying `p`, then `pnat.find hp` is the
+smallest positive natural number satisfying `p`. Note that `pnat.find` is protected,
+meaning that you can't just write `find`, even if the `pnat` namespace is open.
+
+The API for `pnat.find` is:
+
+* `pnat.find_spec` is the proof that `pnat.find hp` satisfies `p`.
+* `pnat.find_min` is the proof that if `m < pnat.find hp` then `m` does not satisfy `p`.
+* `pnat.find_min'` is the proof that if `m` does satisfy `p` then `pnat.find hp ≤ m`.
+-/
+protected def find : ℕ+ :=
+⟨pnat.find_aux h, begin
+  obtain ⟨⟨n, hn⟩, h, -⟩ := pnat.find_aux_spec h,
+  rw h,
+  exact hn
+end⟩
+
+protected theorem find_spec : p (pnat.find h) :=
+begin
+  rw pnat.find,
+  obtain ⟨n, h, hp⟩ := pnat.find_aux_spec h,
+  simp [h, hp]
+end
+
+protected theorem find_min : ∀ {m : ℕ+}, m < pnat.find h → ¬p m :=
+begin
+  simp only [pnat.find, ←pnat.coe_lt_coe, pnat.find_aux, mk_coe, nat.lt_find_iff, exists_prop,
+             not_exists, not_and],
+  rintro ⟨m, hm⟩,
+  contrapose!,
+  intro h,
+  refine ⟨m, le_rfl, m.to_pnat', _, _⟩,
+  { simp [hm] },
+  { convert h,
+    simp [subtype.ext_iff, hm] }
```

Comment:
```suggestion
  exact h m le_rfl _ rfl,
```

Reply:
using `to_pnat'` when you already have a proof that `m` is positive sitting around is just making work for yourself.